// react-router-dom/server.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var router = require('@remix-run/router');
var reactRouterDom = require('react-router-dom');

/**
 * A <Router> that may not navigate to any other location. This is useful
 * on the server where there is no stateful UI.
 */
function StaticRouter({
  basename,
  children,
  location: locationProp = "/"
}) {
  if (typeof locationProp === "string") {
    locationProp = reactRouterDom.parsePath(locationProp);
  }

  let action = router.Action.Pop;
  let location = {
    pathname: locationProp.pathname || "/",
    search: locationProp.search || "",
    hash: locationProp.hash || "",
    state: locationProp.state || null,
    key: locationProp.key || "default"
  };
  let staticNavigator = getStatelessNavigator();
  return /*#__PURE__*/React.createElement(reactRouterDom.Router, {
    basename: basename,
    children: children,
    location: location,
    navigationType: action,
    navigator: staticNavigator,
    static: true
  });
}

/**
 * A Data Router that may not navigate to any other location. This is useful
 * on the server where there is no stateful UI.
 */
function DataStaticRouter({
  data,
  location = "/",
  children
}) {
  // Create a router but do not call initialize() so it has no side effects
  // and performs no data fetching
  let staticRouter = router.createRouter({
    history: router.createMemoryHistory({
      initialEntries: [location]
    }),
    routes: reactRouterDom.createRoutesFromChildren(children),
    hydrationData: data
  });
  !staticRouter.state.initialized ? process.env.NODE_ENV !== "production" ? router.invariant(false, "You must provide a complete `data` prop for <DataStaticRouter>") : router.invariant(false) : void 0;
  let staticNavigator = getStatelessNavigator();
  return /*#__PURE__*/React.createElement(reactRouterDom.UNSAFE_DataRouterContext.Provider, {
    value: staticRouter
  }, /*#__PURE__*/React.createElement(reactRouterDom.UNSAFE_DataRouterStateContext.Provider, {
    value: staticRouter.state
  }, /*#__PURE__*/React.createElement(reactRouterDom.Router, {
    location: staticRouter.state.location,
    navigationType: staticRouter.state.historyAction,
    navigator: staticNavigator,
    static: true
  }, /*#__PURE__*/React.createElement(reactRouterDom.Routes, {
    children: children
  }))));
}

function getStatelessNavigator() {
  return {
    createHref(to) {
      return typeof to === "string" ? to : reactRouterDom.createPath(to);
    },

    push(to) {
      throw new Error(`You cannot use navigator.push() on the server because it is a stateless ` + `environment. This error was probably triggered when you did a ` + `\`navigate(${JSON.stringify(to)})\` somewhere in your app.`);
    },

    replace(to) {
      throw new Error(`You cannot use navigator.replace() on the server because it is a stateless ` + `environment. This error was probably triggered when you did a ` + `\`navigate(${JSON.stringify(to)}, { replace: true })\` somewhere ` + `in your app.`);
    },

    go(delta) {
      throw new Error(`You cannot use navigator.go() on the server because it is a stateless ` + `environment. This error was probably triggered when you did a ` + `\`navigate(${delta})\` somewhere in your app.`);
    },

    back() {
      throw new Error(`You cannot use navigator.back() on the server because it is a stateless ` + `environment.`);
    },

    forward() {
      throw new Error(`You cannot use navigator.forward() on the server because it is a stateless ` + `environment.`);
    }

  };
}

exports.DataStaticRouter = DataStaticRouter;
exports.StaticRouter = StaticRouter;
